plugins/dell/fu-plugin-dell.c

```diff
@@ -522,6 +523,114 @@ fu_plugin_get_results (FuPlugin *plugin, FuDevice *device, GError **error)
 	return TRUE;
 }
 
+static void Esys_Finalize_autoptr_cleanup (ESYS_CONTEXT *esys_context)
+{
+	Esys_Finalize (&esys_context);
+}
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ESYS_CONTEXT, Esys_Finalize_autoptr_cleanup)
+
+static gchar *
+fu_plugin_dell_get_tpm_capabilitity (ESYS_CONTEXT *ctx, guint32 query, GError **error)
+{
+	TSS2_RC rc;
+	guint32 val;
+	g_autofree gchar *result = NULL;
+	g_autofree TPMS_CAPABILITY_DATA *capability = NULL;
+	rc = Esys_GetCapability (ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+	                         TPM2_CAP_TPM_PROPERTIES, query, 1, NULL, &capability);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+		             "failed to get TPM capability %x", query);
+		return NULL;
+	}
+	val = GUINT32_FROM_BE (capability->data.tpmProperties.tpmProperty[0].value);
+	result = g_strstrip (g_strdup ((gchar *)&val));
+	for (guint i = 0; i < strlen(result); i++) {
+		if (!g_ascii_isgraph (result[i]) && result[i] != '\0')
+			result[i] = '\0';
+	}
+
+	return g_steal_pointer (&result);
+}
+
+static gboolean
+fu_plugin_dell_add_tpm_model (FuDevice *dev, GError **error)
+{
+	TSS2_RC rc;
+	const gchar *base = "DELL-TPM";
+	g_autoptr(ESYS_CONTEXT) ctx = NULL;
+	g_autofree gchar *family = NULL;
+	g_autofree gchar *vendor1 = NULL;
+	g_autofree gchar *vendor2 = NULL;
+	g_autofree gchar *vendor3 = NULL;
+	g_autofree gchar *vendor4 = NULL;
+	g_autofree gchar *v1 = NULL;
+	g_autofree gchar *v1_v2 = NULL;
+	g_autofree gchar *v1_v2_v3 = NULL;
+	g_autofree gchar *v1_v2_v3_v4 = NULL;
+
+	rc = Esys_Initialize (&ctx, NULL, NULL);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error_literal (error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND,
+		                     "failed to initialize TPM library");
+		return FALSE;
+	}
+	rc = Esys_Startup (ctx, TPM2_SU_CLEAR);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+		                     "failed to initialize TPM");
+		return FALSE;
+	}
+
+	/* lookup details from TPM */
+	family = fu_plugin_dell_get_tpm_capabilitity (ctx,
+						      TPM2_PT_FAMILY_INDICATOR,
+						      error);
+	if (family == NULL)
+		return FALSE;
+	vendor1 = fu_plugin_dell_get_tpm_capabilitity (ctx,
+						       TPM2_PT_VENDOR_STRING_1,
+						       error);
+	if (vendor1 == NULL)
+		return FALSE;
+	vendor2 = fu_plugin_dell_get_tpm_capabilitity (ctx,
+						       TPM2_PT_VENDOR_STRING_2,
+						       error);
+	if (vendor2 == NULL)
```

Comment:
The properties `TPM2_PT_VENDOR_STRING_2`, `TPM2_PT_VENDOR_STRING_3` and `TPM2_PT_VENDOR_STRING_4` do not need to exist, see the [TPM 2.0 specification Part 3, Section 30.2.1](https://trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-3-Commands-01.38.pdf#page=355):

> NOTE 4 If the vendor string does not require one of these values, the property type does not need to exist.

Therefore these NULL checks shouldn't necessarily be hard failures, although all TPM implementations I know indeed have all four properties even if they are empty.

Reply:
OK